### ISSUE TITLE
figure out how to handle record keeping for CoC violations

### ISSUE THREAD

**Issue by lsmith77** [symfony/diversity#24]
this relates to #1 and #2 

we need some way to keep records of CoC violations.

f.e. to be able to establish a possible pattern of misbehavior or in very extreme cases be able to track a ban, to ensure accountability of the enforcement team and also to be able to learn.

data needs to be stored
* securely
* access needs to be granted explicitly (ie. need to know basis)

This seems like a good use case for keybase.io (https://mobile.twitter.com/lsmith/status/957569554540105729)

potentially data also needs to be stored in different levels of detail. for example we might on the on side keep detailed records of an incident with only minimal access. while we keep a high level summary (like “warning was given to person X, see incident ID xzy for details) to be able to easily validate an attandance list against previous bans etc.

one challenge here is how to deal with European privacy laws that require consent for storing personal data and that constent can be revoked at any point. One way to discourage people from revoking constent might be to in turn remove their public profile as much as possible from the community, like removing blog posts, conf talk videos etc. (https://mobile.twitter.com/lsmith/status/957570035949727744).

**Comment by lsmith77** [comment]
if records are kept in personal keybase accounts. I wonder if EU privacy laws even apply.

**Comment by lsmith77** [comment]
my main concern is records keeping among the people involved in incident reporting and enforcement of future events: ie. enforcement team handling the incident, reporter, accuser, accused but also at least in aggregated form people at future events need access. lets say a very severe case lead to a ban of a specific person. or there is a pattern by one person to be involved in incidents, then managing to down play them sufficiently for no action to happen but if it happens frequently enough one might take a closer look again. this requires some level of long term (of course one could discuss if certain offenses should have a max lifetime in the records).

again I envision that the actual details are kept by the only the enforcement team that handled a given incident .. but that they then forward a “higher level” summary to some aggregated list which all current and future enforcement team members get access to.

ie. so that a conf organizers can name an enforcement team, if the symfony enforcement team agrees, they get access to the list in order to ensure that nobody that was banned is on the attendance list, let alone speaker list .. and that in case they are handling a report, that they can check if the people in that incident have been involved in previous incidents that might show a particular pattern.

**Comment by lsmith77** [comment]
On this topic:
* we do need to explicitly collect consent
* we need to clearly define the purpose for the data we keep and ensure that we keep only a minimal amount of data (this should go into the CoC, ie. #1)
* we do need to deal with people revoking consent however in this case if removing data will cause damage to the public, we can decide to not remove certain data (ie. if someone has several records of sexual harassment we can say that removing those records endangers the public since it will limit our ability to take appropriate actions to ensure safety)

Now the problem is that neither github nor slack provide us with ways to collect this consent.

What I propose is the following:
* we create a database where people can record if they consent to the CoC along with their handles on github/slack/sensioconnect
* we create a solution to enable people to check if someone has consented to the CoC or not for all those handles
* we adapt one of the CLA github bots to add reactions emoticons to all posts that either have consented to the CoC or not (not sure which way around make more sense ..)
* we update the website to show in a comment if someone has consented to the CoC
* we clarify that if anyone takes offense by a statement made on slack, github or the website by someone that has not consented to the CoC, we will remove that content without discussion

**Comment by javiereguiluz** [comment]
All this looks too complicate to me, so I propose to wait and see. Other comparable projects would have to comply with GDPR too, so we can learn from them.

Also, we'd need advice from legal people or links to the source of the GDPR regulation to assure that everything we do is really a legal requirement. Imagine that there's a provision in the regulation to exempt non-profit organizations like ours from implementing some things. That's why we need legal advice, if possible from someone who understands what open source is. Thanks!

**Comment by magnusnordlander** [comment]
Getting some actual legal advice (from actual lawyers specializing in this kind of thing) would definitely be a good idea. If someone knows a lawyer willing to give some advice pro bono, that would be great, otherwise, maybe we can try to get companies to kick in some money (in exchange for something of some value, so that it’s deductible as a business expense) to do a GDPR audit of the entire project?

**Comment by ElectricMaxxx** [comment]
I have an idea who would be the best for that job. I will try to contact him.

**Comment by GawainLynch** [comment]
@javiereguiluz We have given a report on this to the Diversity Team internally. Some of the relevant information, however …

- There is no explicit provision to exempt not-for-profits under GDPR

There are provisions for religious organisations, but that is almost certainly out of scope here.

Also, I would point to Recital 4 of the GDPR:
> The right to the protection of personal data is not an absolute right; it must be considered in relation to its function in society and be balanced against other fundamental rights, in accordance with the principle of proportionality.

- Consulting a lawyer is always useful

Although currently lawyers may be limited in their ability to give accurate advice since it is not yet known precisely how the regulators in different jurisdictions will enforce GDPR.

- Waiting and seeing is a dangerous strategy

Regulators have advised us explicitly saying that they will be unlikely to fine companies that breach the regulation **IF** they can show that they attempted to comply with the regulation to the best of their ability and followed "best practices".

I have to stress here that _"wait and see"_ is the equivalent of burying one's head in the sand and not a wise option.

### Useful Links

- EU GDPR Portal - https://www.eugdpr.org/
- EU GDPR Key Changes - https://www.eugdpr.org/key-changes.html
- EU GDPR FAQs - https://www.eugdpr.org/gdpr-faqs.html
- GDPR Legislation - http://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32016R0679
- GDPR glossary - https://www.eugdpr.org/glossary-of-terms.html

_Warning: the Regulation is a **big** reading exercise_

**Comment by lsmith77** [comment]
note @GawainLynch and a co-worker of his spend a lot of time to dig into all of this which is very much appreciated. I can understand the sentiment of “lets wait and see” but I think this is actually a huge opportunity for us to show how serious we are about CoC enforcement and privacy if we lead the way here.

There are quite a few  CLA bots on which this could be based on one of them.

Alternatively this might be a task for https://github.com/carsonbot/carsonbot ?

/cc @weaverryan

**Comment by hanken** [comment]
In my job I'm currently dealing with GDPR related topics and we've gotten counselling from a data privacy expert (not a lawyer though). I wanted to quickly summarise the advise we got. 

Don't wait and see
We got exactly the same statement as @GawainLynch that authorities will likely not be able to fine you as long you show that you care. 

The minimum that should be done, by the 25.5 is

1. Institute responsible person(s)
Publicly name a responsible person or body. 

2. Create a process inventory
List the data and the intended usage of it (processes). 
Can be as simple as a spreadsheet. 

3. Get some sort of consent. 

In addition I wanted to stress again that the GDPR only applies to person identifiable data (In all the discussions I had, this quickly got forgotten). 

In the case of a contributor commenting on a GitHub issue, the comment itself likely does not fall under the GDPR. The username might does. But then here I would argue that the responsibility is with GitHub. 

Similarly a record of a CoC violation itself is not protected by the GDPR unless the violator is identifiable. So the violations could essentially be public, but without names and only enforcement teams have access to the names.

**Comment by hanken** [comment]
@javiereguiluz I don't there is such a list.
In the GDPR Legislation (http://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32016R0679) Article 4. Paragraph 1 states:

> ‘personal data’ means any information relating to an identified or identifiable natural person (‘data subject’); an identifiable natural person is one who can be identified, directly or indirectly, in particular by reference to an identifier such as a name, an identification number, location data, an online identifier or to one or more factors specific to the physical, physiological, genetic, mental, economic, cultural or social identity of that natural person;

**Comment by nicolas-grekas** [comment]
IMHO, we cannot create *a priori* a database of consents. That's just not proportional to the goal to me.
Instead:
- when a violation is added to the database we'll maintain, we should just ask the parties if they agree that their name is kept inside, with all the confidentiality required
- repeated violations will still be trackable, because we have a brain, and what's inside of it is not covered by GDRP. So we just need to ensure people in the enforcement group talk to each others and maintain some level of knowledge between them. The day this doesn't scale, it'll mean we'ell have too many violations for a reasonable brain. I hope this will never happen. And if it does, we can adapt.
- we need to keep a track record of peoples banned from our online and physical spaces. The reason why doesn't need to be in this track record. This should be allowed by GDPR.

Let's KISS.

**Comment by lsmith77** [comment]
I don’t think putting records into the heads of people is kiss .. its much rather a big risk that this will lead to differences in memory, that repeat violations will not be handled appropriately etc. At any rate, I do acknowledge that asking for consent a priori is a significant intrusion. However to be it does have the advantage that everyone in the community will be explicitly encouraged to look over the CoC at least once.

**Comment by fabpot** [comment]
I'm lost here. How can we enforce "everyone in the community will be explicitly encouraged to look over the CoC at least once"? There is no central place where people "joins" the community, and no requirement for people to sign up on Connect.

One can use Composer to use Symfony without interacting with the rest of the ecosystem. Or do we "only" enforce the CoC when people interact with the community on Github/Slack/conferences? But then, do we also cover Twitter? ... and many other places where people talk about Symfony? I think that's that the piece I don't understand well: where can/do we enforce the CoC?

**Comment by nicolas-grekas** [comment]
On my side, I don't understand how consent give us anything useful: per GDPR, consent can be revoked. Let's say I give it before hand. The day one is accused of anything, one would revoke their consent. Basic protection measure to me. So back to beginning.

**Comment by lsmith77** [comment]
@fabpot indeed we need not worry about people downloading the code. the concern is the interaction within the community. the CoC as proposed does not cover personal twitter accounts, blogs but all interactions on “our” official platforms: github, website, slack. 

@nicolas-grekas there is a provision within the GDPR which allows us to retain data if its in the public interest to for example ensure safety which we think would apply.

**Comment by nicolas-grekas** [comment]
> there is a provision within the GDPR which allows us to retain data if its in the public interest to for example ensure safety which we think would apply

But doesn't this provision apply also without prior consent? Otherwise it's strange:
- I give consent => I cannot be removed per the provision
- I never give consent => I'm "safe" from any safety provision

Do you have some reference about this?

**Comment by magnusnordlander** [comment]
@nicolas-grekas: No, "public interest" is it's own lawful basis for processing personal data. Consent is a separate lawful basis. 

I'm not entirely sure public interest applies, the way I understood it I thought that was mostly for government entities and the like, but I could be absolutely wrong. Basically this is one of those things we won't know for sure util it's been tried in court. There's also the "legitimate interest" basis, which possibly could cover such processing, but that depends on how the CJEU interprets that provision.

However for both the "public interest" and the "legitimate interest", data subjects have the right to object to processing.

> (69) Where personal data might lawfully be processed because processing is necessary for the performance of a task carried out in the public interest or in the exercise of official authority vested in the controller, or on grounds of the legitimate interests of a controller or a third party, a data subject should, nevertheless, be entitled to object to the processing of any personal data relating to his or her particular situation. It should be for the controller to demonstrate that its compelling legitimate interest overrides the interests or the fundamental rights and freedoms of the data subject.

Basically, if the data subject objects, the controller must cease processing until it demonstrates that there is a legitimate interest.

**Comment by nicolas-grekas** [comment]
@magnusnordlander thanks, which means I'm back to my [previous proposal](https://github.com/symfony/diversity/issues/24#issuecomment-374301778). Creating a database of consents *a priori* doesn't provide anything for a CoC to my understanding: prior consent can be objected (fortunately), and as such the CoC database holder will be responsible for proving it is legitimate to keep the record (which it is IMHO, but we cannot be our own judges.)

We really need to focus on what and how.
e.g. the database of CoC-related events has to exist. Even anonymous, it will have a great value. The database of banned ppl/usernames *decoupled from any reason* is also quite easy to claim "legitimate". The database of *not-anonymous* events - or the link between bans and events is much more sensitive (and not strictly required for all purposed. Eg as a SFLive organizer, I'd don't need to know *why* someone cannot register.)

**Comment by lsmith77** [comment]
Well I disagree that there is no value in getting consent before hand:
* It shows we take the GDPR and privacy seriously and have made an actual effort to comply with it because that way we would be on safe legal footing either way as long as nobody asks to get their data removed at which point the GDPR might have been more tested and the limits and special cases will have become more clear to everyone
* It ensures that people are actually aware of how they should behave and may happen if they don’t behave accordingly (this to me has a very significant value)

I actually think the second point is quite significant. Lets not see the GDPR as a law we need to circumvent. It was created for very sensible reasons, to ensure that people know what is done with their data. In other words, to me asking for consent is not a bug its a feature. Because once we do ask for consent everyone knows they should be aware of the CoC and we also know that we can act swiftly in cases of issues: if the person hasn’t accepted the CoC we will just remove the offending content with little debate at all. if the person has accepted the CoC, they cannot claim to not have known that such a thing existed.

But to make it clear, I am willing to skip thinking about the GDPR if the alternative means delaying the CoC adoption.

**Comment by nicolas-grekas** [comment]
I didn't mean to say there is no value in getting consent before, sorry I expressed myself in a confusing way.

With my French culture, proving "good intentions" by being extra conscientious (getting consent when we might find a way without) doesn't look much positive. Seeking for the absolute minimalist process that *does* provide the safety we want to enforce look better.

I think we should really focus on designing processes for what is *strictly* required for having a working CoC enforcement.

Note also that I don't link CoC and GDPR as strongly as you seem to do. CoC doesn't need prior approval to be enforced, exactly like when you enter any public park, there are some rules on the wall I didn't sign off, but still allow the authorities of the park to exclude me when I behave in violation of the rules (which I can then challenge, etc.). I'm *not* saying prior approval wouldn't be better. I just want to way that its should not be a requirement to enforcing the CoC. As such we should design the core processes around this fact. *Then*, when we can seek for prior approval (like when registering for a conference), it will be even easier to enforce the CoC. That's how I reason about it :)

**Comment by ElectricMaxxx** [comment]
I have one question before continue discussion of a database, with all the legal issues: do we currently have a countable amont of violations against a actual version of our CoC? Just to get a feeling about the number we are speaking about. 
Cause the proposed database really sounds like a organisation we call "Schufa" in germany, where each of your debts and violations in paying back your debts are persisted. And just that comparison in my head makes it hard for me to belive in it, to see it getting compatible with law and a compatible with a Community.

**Comment by ElectricMaxxx** [comment]
I mean we must have an image of correct behavior, we should write it down and commit on it, but it is more important to live and show that behavior than tracking that people, who doesn't. 
That would mean:
- offensive reaction on every violation on every channel. 
- consequent and direct ban from a conference, channel, group
 But also:
- being kindful and friendly to guests in our Community
- ...

**Comment by ciaranmcnulty** [comment]
I have found the ICO guide to GDPR the most human-readable so far: https://ico.org.uk/for-organisations/guide-to-the-general-data-protection-regulation-gdpr/lawful-basis-for-processing/

I'm not sure that CoCs fit into any of the other baseis aside from consent. If you are tracking _illegal_ behaviour then it should be forwarded to the authorities; any other records you're keeping are probably going to be subject to consent that can be withdrawn.

I'm not sure how necessary it is to take responsibility for centrally tracking incidents though? Can't a complaint include lists of previous incidents (making it not the project's responsibility to collate things)?

**Comment by lsmith77** [comment]
in terms of GDPR we will base our process on the point that @GawainLynch said about "balancing rights". in terms of record keeping we have a temporary solution in place by the CARE team and in the long run https://www.coc-beacon.com/ seems promising.
